File: src/services/scrapers/genericScraper.js
const axios = require('axios');
const cheerio = require('cheerio');
const config = require('../../config/scraperConfig');
const logger = require('../../utils/logger');

const genericScraper = {
  async scrape(url) {
    try {
      const response = await axios.get(url, {
        headers: {
          'User-Agent': config.scraping.userAgent,
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8'
        },
        timeout: config.scraping.timeout,
        maxRedirects: 5
      });

      const $ = cheerio.load(response.data);
      
      $('script, style, nav, header, footer, .ad, .advertisement').remove();
      
      const title = this.extractTitle($);
      const content = this.extractContent($);
      const comments = this.extractComments($);
      const author = this.extractAuthor($);
      const timestamp = this.extractTimestamp($);
      const metadata = this.extractMetadata($);

      return {
        post_text: title,
        content: content,
        comments: comments,
        images: [],
        links: [],
        timestamp: timestamp,
        author: author,
        likes: 0,
        shares: 0,
        engagement: {
          likes: 0,
          comments: comments.length,
          shares: 0
        },
        metadata: {
          ...metadata,
          word_count: content.split(' ').length,
          url: url
        }
      };
    } catch (error) {
      logger.error(`Generic scraping error for ${url}: ${error.message}`);
      return { error: error.message };
    }
  },

  extractTitle($) {
    const titleSelectors = [
      'h1',
      'meta[property="og:title"]',
      'meta[name="twitter:title"]',
      'title',
      '.post-title',
      '.article-title',
      '[class*="title"]'
    ];
    
    for (const selector of titleSelectors) {
      const el = $(selector).first();
      if (el.length) {
        const text = el.attr('content') || el.text().trim();
        if (text && text.length > 5) {
          return text;
        }
      }
    }
    
    return 'Untitled';
  },

  extractContent($) {
    const contentSelectors = [
      'article',
      '.post-content',
      '.entry-content',
      '.article-content',
      '.content',
      'main',
      '[role="main"]'
    ];
    
    let content = '';
    
    for (const selector of contentSelectors) {
      const container = $(selector).first();
      if (container.length) {
        const paragraphs = [];
        container.find('p').each((i, el) => {
          const text = $(el).text().trim();
          if (text.length > 50) {
            paragraphs.push(text);
          }
        });
        
        if (paragraphs.length > 0) {
          content = paragraphs.join('\n');
          break;
        }
      }
    }
    
    if (!content) {
      const paragraphs = [];
      $('p').each((i, el) => {
        const text = $(el).text().trim();
        if (text.length > 50) {
          paragraphs.push(text);
        }
      });
      content = paragraphs.slice(0, 20).join('\n');
    }
    
    return content || 'No content extracted';
  },

  extractComments($) {
    const comments = [];
    const commentSelectors = [
      '.comment',
      '.comment-item',
      '.comment-body',
      '[class*="comment"]',
      '[id*="comment"]',
      '.discussion-item',
      '.reply'
    ];
    
    const seenTexts = new Set();
    
    commentSelectors.forEach(selector => {
      $(selector).each((i, el) => {
        const $el = $(el);
        const text = $el.text().trim();
        
        if (text.length > 20 && text.length < 2000 && !seenTexts.has(text)) {
          seenTexts.add(text);
          
          const authorEl = $el.find('.author, .comment-author, [class*="author"]').first();
          const author = authorEl.text().trim() || 'Anonymous';
          
          const dateEl = $el.find('time, .date, [class*="date"]').first();
          const date = dateEl.attr('datetime') || dateEl.text().trim();
          
          comments.push({
            user: author,
            text: text,
            likes: 0,
            timestamp: date ? this.parseDate(date) : new Date().toISOString(),
            replies_count: 0
          });
        }
      });
    });

    return comments.slice(0, 100);
  },

  extractAuthor($) {
    const authorSelectors = [
      'meta[name="author"]',
      'meta[property="article:author"]',
      '[rel="author"]',
      '.author',
      '.author-name',
      '[class*="author"]',
      '[itemprop="author"]'
    ];
    
    for (const selector of authorSelectors) {
      const el = $(selector).first();
      if (el.length) {
        const author = el.attr('content') || el.text().trim();
        if (author && author.length > 0 && author.length < 100) {
          return author;
        }
      }
    }
    
    return 'Unknown';
  },

  extractTimestamp($) {
    const dateSelectors = [
      'time[datetime]',
      'meta[property="article:published_time"]',
      'meta[name="publish-date"]',
      '.publish-date',
      '.post-date',
      '[class*="date"]',
      '[itemprop="datePublished"]'
    ];
    
    for (const selector of dateSelectors) {
      const el = $(selector).first();
      if (el.length) {
        const date = el.attr('datetime') || el.attr('content') || el.text().trim();
        if (date) {
          try {
            return new Date(date).toISOString();
          } catch (e) {
            continue;
          }
        }
      }
    }
    
    const bodyText = $('body').text();
    const datePatterns = [
      /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/,
      /\d{4}-\d{2}-\d{2}/,
      /\d{2}\/\d{2}\/\d{4}/,
      /(January|February|March|April|May|June|July|August|September|October|November|December)\s+\d{1,2},?\s+\d{4}/i
    ];
    
    for (const pattern of datePatterns) {
      const match = bodyText.match(pattern);
      if (match) {
        try {
          return new Date(match[0]).toISOString();
        } catch (e) {
          continue;
        }
      }
    }
    
    return new Date().toISOString();
  },

  extractMetadata($) {
    const metadata = {};
    
    metadata.description = $('meta[name="description"]').attr('content') || '';
    metadata.keywords = $('meta[name="keywords"]').attr('content') || '';
    metadata.language = $('html').attr('lang') || 'en';
    
    const ogTitle = $('meta[property="og:title"]').attr('content');
    const ogDescription = $('meta[property="og:description"]').attr('content');
    const ogImage = $('meta[property="og:image"]').attr('content');
    
    if (ogTitle) metadata.og_title = ogTitle;
    if (ogDescription) metadata.og_description = ogDescription;
    if (ogImage) metadata.og_image = ogImage;
    
    return metadata;
  },

  parseDate(dateString) {
    try {
      const date = new Date(dateString);
      if (!isNaN(date.getTime())) {
        return date.toISOString();
      }
    } catch (e) {
    }
    return new Date().toISOString();
  }
};

module.exports = genericScraper;

File: python_scrapers/scrapers/generic_scraper.py
import requests
from bs4 import BeautifulSoup
from datetime import datetime
import re

class GenericScraper:
    def __init__(self):
        self.headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        }
    
    def scrape_post(self, url):
        try:
            response = requests.get(url, headers=self.headers, timeout=30)
            response.raise_for_status()
            
            soup = BeautifulSoup(response.content, 'html.parser')
            
            title = self._extract_title(soup)
            content = self._extract_content(soup)
            comments = self._extract_comments(soup)
            author = self._extract_author(soup)
            timestamp = self._extract_timestamp(soup)
            
            return {
                'url': url,
                'post_text': title + '\n' + content,
                'author': author,
                'comments': comments,
                'likes': 0,
                'shares': 0,
                'timestamp': timestamp,
                'post_type': 'article'
            }
        
        except Exception as e:
            return {'error': f'Generic scraping failed: {str(e)}'}
    
    def scrape_profile(self, url):
        return {'error': 'Profile scraping not supported for generic URLs'}
    
    def _extract_title(self, soup):
        title_tag = soup.find('h1')
        if title_tag:
            return title_tag.get_text(strip=True)
        
        title_tag = soup.find('title')
        if title_tag:
            return title_tag.get_text(strip=True)
        
        return ''
    
    def _extract_content(self, soup):
        paragraphs = []
        for p in soup.find_all('p'):
            text = p.get_text(strip=True)
            if len(text) > 20:
                paragraphs.append(text)
        
        return '\n'.join(paragraphs[:10])
    
    def _extract_comments(self, soup):
        comments = []
        comment_selectors = [
            {'class': re.compile(r'comment', re.I)},
            {'id': re.compile(r'comment', re.I)}
        ]
        
        for selector in comment_selectors:
            comment_elements = soup.find_all(attrs=selector)
            for el in comment_elements[:30]:
                text = el.get_text(strip=True)
                if 10 < len(text) < 1000:
                    comments.append({
                        'user': 'Anonymous',
                        'text': text,
                        'likes': 0,
                        'timestamp': datetime.now().isoformat()
                    })
        
        return comments[:50]
    
    def _extract_author(self, soup):
        author_meta = soup.find('meta', attrs={'name': 'author'})
        if author_meta and author_meta.get('content'):
            return author_meta['content']
        
        author_selectors = [
            {'class': re.compile(r'author', re.I)},
            {'rel': 'author'}
        ]
        
        for selector in author_selectors:
            author_el = soup.find(attrs=selector)
            if author_el:
                text = author_el.get_text(strip=True)
                if text:
                    return text
        
        return 'Unknown'
    
    def _extract_timestamp(self, soup):
        time_tag = soup.find('time')
        if time_tag and time_tag.get('datetime'):
            return time_tag['datetime']
        
        date_patterns = [
            r'\d{4}-\d{2}-\d{2}',
            r'\d{2}/\d{2}/\d{4}'
        ]
        
        body_text = soup.get_text()
        for pattern in date_patterns:
            match = re.search(pattern, body_text)
            if match:
                try:
                    return datetime.fromisoformat(match.group(0)).isoformat()
                except:
                    pass
        
        return datetime.now().isoformat()